src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -110,73 +117,97 @@ void Run (DirectoryAssemblyResolver res)
 			}
 
 			// Put every assembly we'll need in the resolver
+			bool hasExportReference = false;
+			bool haveMonoAndroid = false;
+			var allTypemapAssemblies = new HashSet<string> (StringComparer.OrdinalIgnoreCase);
+			var userAssemblies = new Dictionary<string, string> (StringComparer.OrdinalIgnoreCase);
 			foreach (var assembly in ResolvedAssemblies) {
-				if (bool.TryParse (assembly.GetMetadata (AndroidSkipJavaStubGeneration), out bool value) && value) {
+				bool value;
+				if (bool.TryParse (assembly.GetMetadata (AndroidSkipJavaStubGeneration), out value) && value) {
 					Log.LogDebugMessage ($"Skipping Java Stub Generation for {assembly.ItemSpec}");
 					continue;
 				}
+
+				bool addAssembly = false;
+				string fileName = Path.GetFileName (assembly.ItemSpec);
+				if (!hasExportReference && String.Compare ("Mono.Android.Export.dll", fileName, StringComparison.OrdinalIgnoreCase) == 0) {
+					hasExportReference = true;
+					addAssembly = true;
+				} else if (!haveMonoAndroid && String.Compare ("Mono.Android.dll", fileName, StringComparison.OrdinalIgnoreCase) == 0) {
+					haveMonoAndroid = true;
+					addAssembly = true;
+				} else if (MonoAndroidHelper.FrameworkAssembliesToTreatAsUserAssemblies.Contains (fileName)) {
+					if (!bool.TryParse (assembly.GetMetadata (AndroidSkipJavaStubGeneration), out value) || !value) {
+						string name = Path.GetFileNameWithoutExtension (fileName);
+						if (!userAssemblies.ContainsKey (name))
+							userAssemblies.Add (name, assembly.ItemSpec);
+						addAssembly = true;
+					}
+				}
+
+				if (addAssembly) {
+					if (!allTypemapAssemblies.Contains (assembly.ItemSpec))
```

Comment:
`allTypemapAssemblies` is a `HashSet<string>`; why do we need this `.Contains()` call in the first place?  Skip it, and just use `.Add()`.